src/tax/basis.js

```diff
@@ -0,0 +1,66 @@
+// @flow
+
+import moment from "moment";
+
+export type BasisFrame = {
+  date: moment,
+  amount: number,
+  cost: number,
+};
+
+/** A cost basis calculator with LIFO (Last In, First Out) semantics.
+ * Any time that a currency is being disposed of (e.g. sold or gifted),
+ * it reports the effective cost basis for that amount of currency, by
+ * returning the appropriate BasisFrames. Note that it may return several
+ * BasisFrames, e.g. if you acquired 50 at one cost, 50 at another cost, and then
+ * dispose of 75, you will get 2 frames (50 from the last frame, 25 from the first one).
+ */
+export class LIFOCostBasisCalculator {
+  frames: BasisFrame[];
+  ticker: string;
+
+  constructor(ticker: string) {
+    this.frames = [];
+    this.ticker = ticker;
+  }
+
+  /** Record that a cryptocurrency was acquired at a certain cost */
+  acquire(date: moment, amount: number, cost: number) {
+    const frame = {amount, cost, date};
```

Comment:
Suggest throwing a hard error here unless `amount` is an integer no greater than `Number.MAX_SAFE_INTEGER`. You should assert similarly at each point where you take a `number` as input.

You never want to use a floating-point type for money, because you get things like `0.1 + 0.2 !== 0.3`. In JavaScript, the only numeric type is a double-precision float, which is a problem. As long as your values and all intermediate computations stay fully representable in the mantissa, you _should_ be fine.

(Honestly, if I were writing this, I would consider skipping JavaScript for the backend (read: any logic) and using a language that enables me to require that overflow throw an error, such as Rust.)

Reply:
Switched to big.js